### DISCUSSION TITLE
兑换失败是什么原因

### DISCUSSION THREAD

**Discussion by kuronekoda** [Ljzd-PRO/Mys_Goods_Tool#136] [Q&A]
![Snipaste_2023-06-30_18-00-12](https://github.com/Ljzd-PRO/Mys_Goods_Tool/assets/136661812/2d87d7a2-4e17-48fa-a53b-ef5c4182358f)
06-30 17:58:34 [INFO] mys_goods_tool.exchange_mode | 已添加定时兑换任务 74510315 - 【崩坏：星穹铁道】鼠标垫-星旅一瞬 - 2023-06-30 18:00:00
06-30 18:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 74510315 商品 2023052611696 兑换失败，可以自行确认。
06-30 18:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
06-30 18:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 74510315 - 【崩坏：星穹铁道】鼠标垫-星旅一瞬 - 线程 1 - 兑换失败
06-30 18:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 74510315 商品 2023052611696 兑换失败，可以自行确认。
06-30 18:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
06-30 18:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 74510315 - 【崩坏：星穹铁道】鼠标垫-星旅一瞬 - 线程 2 - 兑换失败

**Comment by kuronekoda** [comment]
貌似和[#121 ](url)一个原因，已经这样兑换失败三次了，真的很喜欢这款鼠标垫，希望作者可以修复一下，十分感谢！

**Comment by Ljzd-PRO** [comment]
> 貌似和[#121 ](url)一个原因，已经这样兑换失败三次了，真的很喜欢这款鼠标垫，希望作者可以修复一下，十分感谢！

是的

**Comment by Ljzd-PRO** [comment]
又尝试修复了一下，可以试一下：https://github.com/Ljzd-PRO/Mys_Goods_Tool/actions/runs/5479310149

解决了几个可能的问题：
- `device_config`, `salt_config` 如果程序默认值更新了但是用户数据文件用的还是旧的，旧的配置还是会覆盖掉默认值的问题
- 兑换请求的Headers里增加`Referer`，修改`Origin`, `Content-Type`
- Headers的`User-Agent`里的App版本号部分和`x-rpc-app_version`不一致的问题

除了Headers里没有`x-rpc-verify_key`, `x-rpc-device_fp`，其他应该都和米游社App里实际发出的请求一致了。

**Comment by ayakasuki** [comment]
我也是 这次兑换了3次
![image](https://github.com/Ljzd-PRO/Mys_Goods_Tool/assets/20859971/47e22b23-6283-4363-afe4-8376f56f3411)
